salt/tops/varstack.py

```diff
@@ -45,28 +45,14 @@
 
 from __future__ import absolute_import, print_function, unicode_literals
 
-# Import python libs
-import logging
-
 try:
-    HAS_VARSTACK = False
     import varstack
-    HAS_VARSTACK = True
 except ImportError:
-    pass
-
-# Set up logging
-log = logging.getLogger(__name__)
-
-# Define the module's virtual name
-__virtualname__ = 'varstack'
+    varstack = None
 
 
 def __virtual__():
-    if not HAS_VARSTACK:
-        log.error("Can't find varstack master_top")
-        return False
-    return __virtualname__
+    return bool(varstack), 'The varstack module could not be loaded: varstack dependency is missing.'
```

Comment:
This will need to return the virtual name when renamed.

Reply:
Technically that is not true in this case. When the `__virtual__` returns `True`, the loader will fall back to the module name (i.e. without the .py). In this case, since it is called `varstack.py`, it would be loaded as `varstack`. So, it would work, though via a happy accident.

So it should still be changed to something like this:

```python
if varstack is None:
    return (False, 'The varstack module could not be loaded: varstack dependency is missing.')
return __virtualname__
```

Note that the same change should be made to the external pillar module.

Reply:
apparently my other comment didn't get posted.

I was also requesting that varstack.py gets renamed back to varstack_pillar.py, so that we do not run into the same problems we have seen with boto/boto3 and docker -> dockermod

Reply:
But is really odd. Sounds to me like PHP problems: everything in one pile. Loader problems? Python has namespace, why is that now `foo.bar.spam` needs to clash with `foo.fred.spam`?

Reply:
Also, apparently I jumped over to the varstack tops file when reviewing and didn't realize it.

Reply:
I am just trying to avoid another problem like this https://github.com/saltstack/salt/pull/47726

Reply:
But we should be able to have the same package names in different namespace. To me it sounds like a flaw in the Loader. Because renaming all around the place different parts into different names in order let them do not clash together sounds like a workaround, not really a fix.

Maybe we can leave this PR as is and then look into this topic?

Reply:
@isbm Our loader temporarily monkeypatches `sys.path` while loading modules to include the directory containing the modules which are loaded. So, if the module is named `foo.py` and it does an `import foo` within it, then it will try to load itself instead of `/usr/lib/python3.6/site-packages/foo/__init__.py` and that will cause problems.

Reply:
it isn't a problem with namespaces for modules.

It is a problem that we are doing `import varstack` inside this module.

https://github.com/saltstack/salt/pull/48436/files#diff-caa464e4e9ee6e852e118724583cc313L29

Reply:
@terminalmage yeah, in that way it will crash for sure, as it doesn't use the namespace. How we can fix this (without randomising names of the components)? Enforce full import? Reject modules that are so? Other ideas? Because it is nicer to have `salt.modules.docker` and `salt.pillar.docker` instead of `salt.modules.dockermodule` and `salt.pillar.docker_this_time_a_pillar`.

Reply:
@gtmanfred well, then just fix the import?

Reply:
> So it should still be changed to something like this:

```
if varstack is None:
    return (False, 'The varstack module could not be loaded: varstack dependency is missing.')
return __virtualname__
```

@terminalmage  @gtmanfred This is basically what I did in my initial PR, but then changed it based on the feedback.

If necessary, I can resubmit my initial PR which was more idiomatic, as far as I can tell by looking at existing modules.

Reply:
@max-arnold OK, the loader seem has a problems that we need to tackle them later. The way it is now "fixed" is just that you don't have the same name in the module/pillars. And so when we have `varstack` in pillars, `varstack` in modules and `varstack` in Python, then things gets messy.

In that regard, lets do the following:

- Rename (or leave?) `salt.pillars.varstack_pillar`
- Rename `salt.modules.varstackmod`
- `varstack` — comes from the Python

```python
__virtualname__ = `varstack`

def __virtual__():
    return (varstack and __virtualname__ or False,
            'The varstack module could not be loaded: varstack dependency is missing.')
```

Then this should work. In a long run there should be either rules of the naming of the modules/pillars/states or a way to stop these sort of clashes.